File: cmd/agenthub-gui/frontend/src/pages/events.ts
// Events page: the control-plane timeline.
//
// It answers the question the Servers page cannot: not "is this server
// healthy" but "how did it get that way". The two belong together and are
// deliberately not merged — a status is a value, a history is a sequence,
// and a page that tries to be both ends up refreshing one of them wrongly.
//
// Read-only, like Sessions and for a stronger reason: these records describe
// things that already happened, so there is nothing here to change.
//
// The vocabulary is CLOSED (docs/modules/foundation.md), which is what makes
// the tone mapping below legitimate. Colouring rows by a log message would
// be guessing; colouring them by `kind` is reading a contract. A kind this
// build does not know still renders — it lands in the neutral tone rather
// than being hidden, because a frontend older than its daemon must not
// silently drop records.

import { EVT, hub, on } from "../bridge";
import { clear, el, empty, loadingState, pageHeader, relTime, section, table } from "../dom";
import type { Page } from "../page";
import { failureBox } from "../page";
import { button, selectInput } from "../ui";
import { Pager, filterBar, pagerFooter, rangeField, rangeMillis } from "./observe";
import type { EventLog, EventRecord, TopicEvent } from "../types";

/** How many records one PAGE holds. The stream is low-rate, so this is
 *  usually the whole of an incident-sized window — but "usually" is not a
 *  contract, and a hub that flapped for an hour has more to say than one
 *  screen. The pager is what makes that readable rather than truncated. */
const PAGE = 300;

/** Tone per kind. Anything unlisted is neutral — including a kind from a
 *  newer daemon, which must still be shown. */
const TONES: Record<string, "bad" | "warn" | "good"> = {
  connect_failed: "bad",
  respawn_failed: "bad",
  health_down: "bad",
  secrets_missing: "bad",
  oauth_refresh_failed: "bad",
  registry_reload_failed: "bad",
  circuit_open: "warn",
  circuit_half_open: "warn",
  disconnected: "warn",
  respawned: "warn",
  stopping: "warn",
  stopped: "warn",
  connected: "good",
  health_up: "good",
  circuit_closed: "good",
  started: "good",
};

/** What `count` counts, per kind — the mirror of eventlog.CountNoun, kept
 *  here for the same reason TONES is: this page may not import internal/*,
 *  and the vocabulary it keys on is closed and published. An absent kind
 *  falls back to `n=13`, which is what every count looked like before the
 *  noun existed and is why a connect listing thirteen tools read as a
 *  thirteenth attempt. */
const COUNT_NOUN: Record<string, string> = {
  connected: "tools",
  tools_changed: "tools",
  respawned: "respawns",
  respawn_failed: "respawns",
  disconnected: "reconnects",
  circuit_open: "failures",
  health_down: "failures",
  health_up: "failures",
};

/** serverOf is the downstream a record is about, with its derived instance
 *  when there is one — one identity, so one cell. */
export function serverOf(e: EventRecord): string {
  if (!e.server) return "";
  return e.inst ? `${e.server}/${e.inst}` : e.server;
}

/** actorCell is who ASKED, which is a different question from which server a
 *  record is about and therefore a different column. A record can carry both:
 *  one subject cell picking a winner between them showed a login's server and
 *  never said whose session it was for.
 *
 *  A session is labelled rather than printed bare. The HTTP face's callers are
 *  tokens rather than configured clients, so a session id is all the identity
 *  there is — and an opaque id sitting under a "Client" header reads as a
 *  client whose name happens to look like a hash. */
function actorCell(e: EventRecord): Node {
  if (e.client) {
    return el("div", {}, [
      el("div", { text: e.client }),
      e.session ? el("div", { class: "muted mono", text: `session ${e.session}` }) : null,
    ]);
  }
  if (e.session) {
    return el("div", {}, [
      el("div", { class: "mono", text: e.session }),
      el("div", { class: "muted", text: "session" }),
    ]);
  }
  return el("div", { class: "muted", text: "—" });
}

/** detailOf folds the transition, the count and the free text into one
 *  string, in that order: the flip is the fact, the prose elaborates it. */
export function detailOf(e: EventRecord): string {
  const parts: string[] = [];
  if (e.from || e.to) parts.push(`${e.from || "—"} → ${e.to || "—"}`);
  if (e.count) {
    const noun = COUNT_NOUN[e.kind];
    parts.push(noun ? `${e.count} ${noun}` : `n=${e.count}`);
  }
  if (e.rev) parts.push(`rev ${e.rev}`);
  if (e.durMs) parts.push(`${e.durMs}ms`);
  if (e.detail) parts.push(e.detail);
  return parts.join(" · ");
}

/** The columns eventRows fills, in order. It ships beside the renderer for
 *  the reason the renderer is shared at all: a caller holding its own header
 *  list mislabels every cell under it the day this one moves. */
export const EVENT_COLUMNS = ["Time", "PID", "Scope", "What", "Server", "Client", "Detail"];

/** eventRows renders the shared table body, so the Servers panel and this
 *  page cannot drift into two ways of reading one record.
 *
 *  One value per column, and the three that used to share a cell — the scope,
 *  the server, the caller — each get their own. Stacked, a `daemon` under a
 *  server's name read as a fourth identity of that server rather than as the
 *  kind of process the record came from. */
export function eventRows(events: EventRecord[]): (Node | string)[][] {
  return events.map((e) => [
    el("div", { text: relTime(e.ts) }),
    el("div", { class: "muted mono", text: e.pid ? String(e.pid) : "—" }),
    el("div", { class: "muted", text: e.scope || "—" }),
    el("span", { class: `badge ${TONES[e.kind] ?? "neutral"}`, text: e.kind }),
    el("div", { text: serverOf(e) || "—" }),
    actorCell(e),
    el("div", { class: "muted", text: detailOf(e) || "—" }),
  ]);
}

/** "Only what went wrong" is the daemon's own answer now: every record
 *  carries a CLASS, routine or disruption, derived from its kind. This page
 *  used to assemble that set itself out of TONES, which was a second copy of
 *  a mapping that lives in internal/eventlog — and a copy that could only be
 *  as current as the frontend build.
 *
 *  Note what the class does that a tone cannot: `health_up` and
 *  `circuit_closed` are GOOD news and still belong to the disruption, because
 *  they are how it ended. A filter built from tones dropped them and showed
 *  every outage beginning with none of them finishing. */
const CLASS_DISRUPTION = "disruption";

/** How many records the facet read pulls. Larger than PAGE because it exists
 *  to enumerate what CAN be selected, and a dropdown that omits a server
 *  because it fell off a 300-record tail is a filter that hides data. */
const FACET_PAGE = 2000;

/** What the dropdowns offer, counted. */
interface Facets {
  servers: Record<string, number>;
  clients: Record<string, number>;
  kinds: Record<string, number>;
}

function facetsOf(events: EventRecord[]): Facets {
  const facets: Facets = { servers: {}, clients: {}, kinds: {} };
  for (const e of events) {
    if (e.server) facets.servers[e.server] = (facets.servers[e.server] ?? 0) + 1;
    if (e.client) facets.clients[e.client] = (facets.clients[e.client] ?? 0) + 1;
    if (e.kind) facets.kinds[e.kind] = (facets.kinds[e.kind] ?? 0) + 1;
  }
  return facets;
}

export function eventsPage(): Page {
  let root: HTMLElement | null = null;
  let off: (() => void) | null = null;
  let rangeHours = 24;
  let scope = "";
  let server = "";
  let client = "";
  let cls = "";
  let kind = "";
  let log: EventLog | null = null;
  const pager = new Pager();
  let facets: Facets = { servers: {}, clients: {}, kinds: {} };
  let failure: unknown = null;
  let loading = true;

  const sinceMillis = (): number => rangeMillis(rangeHours);
  const filtered = (): boolean =>
    scope !== "" || server !== "" || client !== "" || cls !== "" || kind !== "";

  /** Every selector goes to the DAEMON, not to the rendered rows.
   *
   *  The read is limited, so filtering a page client-side would search only
   *  the newest PAGE records and report "nothing matches" for something that
   *  is merely older than the window — the two are indistinguishable to the
   *  person reading it, and the wrong one sends them looking for a bug that
   *  is not there. */
  async function load(): Promise<void> {
    try {
      log = await hub.eventLog(sinceMillis(), PAGE, scope, server, client, cls,
        kind ? [kind] : [], pager.current());
      pager.accept(log.nextCursor);
      failure = null;
    } catch (err) {
      failure = err;
    } finally {
      loading = false;
      draw();
    }
  }

  /** Facets come from an UNFILTERED read of the same range, so narrowing by
   *  one selector does not empty the others. Deriving them from the filtered
   *  rows instead is the trap: picking a server would leave that server as
   *  the only option in its own dropdown, with no way back. */
  async function loadFacets(): Promise<void> {
    try {
      const all = await hub.eventLog(sinceMillis(), FACET_PAGE, "", "", "", "", []);
      facets = facetsOf(all.events);
    } catch {
      // A facet read that fails costs the dropdown its options, never the
      // page its rows. The selectors still work; they just stop suggesting.
      facets = { servers: {}, clients: {}, kinds: {} };
    }
  }

  /** Reload rows, and the facets too when the RANGE moved — the facet set is
   *  a function of the range alone. */
  function refresh(rangeMoved = false): void {
    // Back to page one. A cursor taken under one filter names a row the next
    // filter may not contain, and paging on from it would skip records
    // without ever saying so.
    pager.reset();
    loading = true;
    draw();
    void (rangeMoved ? loadFacets().then(load) : load());
  }

  function field(label: string, control: HTMLElement): HTMLElement {
    return el("label", { class: "activity-filter-field" }, [el("span", { text: label }), control]);
  }

  /** One dropdown over a counted facet.
   *
   *  A value that is currently selected but absent from the facets is still
   *  offered, shown as (0): it is reachable through a range change, and
   *  silently dropping it would reset the filter under the operator. */
  function facetField(
    label: string, allLabel: string, current: string,
    values: Record<string, number>, onChange: (v: string) => void,
  ): HTMLElement {
    const options = [{ value: "", label: allLabel }];
    if (current && !(current in values)) options.push({ value: current, label: `${current} (0)` });
    for (const [value, count] of Object.entries(values).sort((a, b) => a[0].localeCompare(b[0]))) {
      options.push({ value, label: `${value} (${count})` });
    }
    const select = selectInput(options, current);
    select.addEventListener("change", () => {
      onChange(select.value);
      refresh();
    });
    return field(label, select);
  }

  function filters(): HTMLElement {
    const scopePick = selectInput(
      [
        { value: "", label: "All scopes" },
        { value: "server", label: "Servers" },
        { value: "gateway", label: "Gateways" },
        { value: "daemon", label: "Daemon" },
      ],
      scope,
    );
    scopePick.addEventListener("change", () => {
      scope = scopePick.value;
      refresh();
    });

    // Class is its own control, above the kinds, because it is the reason
    // most people open this page and because it is not a kind: it is the
    // question "is this the hub working or the hub in trouble", which the
    // daemon answers for every record including ones this build cannot name.
    const classPick = selectInput(
      [
        { value: "", label: "Everything" },
        { value: CLASS_DISRUPTION, label: "⚠ Disruptions only" },
        { value: "routine", label: "Routine only" },
      ],
      cls,
    );
    classPick.addEventListener("change", () => {
      cls = classPick.value;
      refresh();
    });

    const kindOptions = [{ value: "", label: "All kinds" }];
    if (kind && !(kind in facets.kinds)) {
      kindOptions.push({ value: kind, label: `${kind} (0)` });
    }
    for (const [k, count] of Object.entries(facets.kinds).sort((a, b) => a[0].localeCompare(b[0]))) {
      kindOptions.push({ value: k, label: `${k} (${count})` });
    }
    const kindPick = selectInput(kindOptions, kind);
    kindPick.addEventListener("change", () => {
      kind = kindPick.value;
      refresh();
    });

    return filterBar(
      rangeField(rangeHours, (hours) => {
        rangeHours = hours;
        refresh(true);
      }),
      field("Show", classPick),
      field("Scope", scopePick),
      facetField("Server", "All servers", server, facets.servers, (v) => { server = v; }),
      facetField("Client", "All clients", client, facets.clients, (v) => { client = v; }),
      field("Kind", kindPick),
    );
  }

  function clearFilters(): void {
    scope = server = client = cls = kind = "";
    refresh();
  }

  function body(): Node {
    if (loading) return loadingState("Reading the event log…", 6);
    if (failure) return failureBox(failure);
    const events = log?.events ?? [];
    if (events.length === 0) {
      // THREE facts an operator has to be able to tell apart, and only the
      // first means anything is wrong with the setup: nothing has ever been
      // recorded (usually the stream is switched off), the hub was quiet in
      // this range, or the selectors excluded everything.
      if (log && log.files === 0) {
        return empty(
          "Nothing has been recorded yet.",
          "Any running gateway or daemon writes this stream unless events.enabled is false.",
        );
      }
      if (filtered()) {
        return empty(
          "No state changes match these filters.",
          "The range holds records; these selectors exclude them.",
          button("Clear filters", "btn", clearFilters),
        );
      }
      return empty("No state changes in this range.", "The hub has been quiet — which is the good case.");
    }
    // The daemon serves newest first, like every other observability list:
    // an operator opens this page because something just happened.
    return table(EVENT_COLUMNS, eventRows(events), "observe-table");
  }

  function draw(): void {
    if (!root) return;
    const reload = button("Refresh", "btn", () => refresh(true));
    clear(root);
    root.append(
      pageHeader(
        "Events",
        "Every state change of a server, a gateway or the daemon, in one timeline.",
        reload,
      ),
      section("Timeline", filters(), body(), pagerFooter({
        pager,
        pageSize: PAGE,
        shown: log?.events?.length ?? 0,
        total: log?.total ?? 0,
        noun: "events",
        onChange: () => {
          loading = true;
          draw();
          void load();
        },
      })),
    );
  }

  return {
    render(node) {
      root = node;
      // The SSE stream says only THAT something changed; the records are read
      // back through the control plane. Using it as a re-read trigger is
      // exactly the contract internal/event states for the bus.
      off = on<TopicEvent>(EVT.servers, () => void load());
      draw();
      // Facets first, so the dropdowns are populated by the time the rows
      // land rather than filling in a moment later under the pointer.
      return loadFacets().then(load);
    },
    dispose() {
      off?.();
      off = null;
      root = null;
      log = null;
      facets = { servers: {}, clients: {}, kinds: {} };
    },
  };
}

File: internal/cli/events.go
package cli

import (
	"context"
	"fmt"
	"io"
	"path/filepath"

	"strings"
	"text/tabwriter"
	"time"

	"github.com/spf13/cobra"

	"github.com/dinstein/agent-hub/internal/eventlog"
)

// `events` reads the control-plane event stream: what happened to a
// downstream server, a gateway or the daemon, in the closed vocabulary of
// internal/eventlog.
//
// WHAT THIS COMMAND USED TO BE. It subscribed to the daemon's SSE stream,
// was online-only (exit 4 without a daemon), and rendered whatever that
// stream carried — which, per the contract in internal/event, is
// deliberately a change NOTIFICATION channel and NOT a change log: a
// subscriber that misses a frame re-reads state, and the payload is a full
// snapshot rather than a description of what moved. So the old command
// answered "something about servers changed just now" and could not answer
// "what happened to this server", which is the question people brought to
// it.
//
// The event stream IS the change log, so the verb moves to it. Two
// consequences follow, and both are improvements:
//
//   - It works OFFLINE, and must. A stdio gateway writes this file with no
//     daemon anywhere in the picture, so an installation without one is not
//     a degraded case here — it is the ordinary one, and the old exit 4
//     refused exactly the setup with the most to explain.
//   - --follow tails the file rather than holding a subscription, so it
//     survives a daemon restart instead of ending on one.
const (
	// eventsDefaultLimit is how many records a bare `events` shows. The
	// stream is low-rate, but an incident-scale window still wants a tail
	// rather than the whole file.
	eventsDefaultLimit = 100
	// eventsFollowInterval is the --follow re-read period. It matches the
	// other two followers so all three feel the same.
	eventsFollowInterval = 500 * time.Millisecond
	// eventsDetailWidth caps the DETAIL column. The full text is in --json.
	eventsDetailWidth = 90
)

// EventRow is one record as both output modes render it.
type EventRow struct {
	TS     string `json:"ts"`
	Scope  string `json:"scope"`
	Kind   string `json:"kind"`
	Server string `json:"server,omitempty"`
	Inst   string `json:"inst,omitempty"`
	Client string `json:"client,omitempty"`
	PID    int    `json:"pid,omitempty"`
	From   string `json:"from,omitempty"`
	To     string `json:"to,omitempty"`
	Detail string `json:"detail,omitempty"`
	// Count is the one number the kind carries; eventlog.CountNoun says what
	// it counts. The record has one such field and so does its projection.
	Count int `json:"count,omitempty"`
	// Rev is a config generation, not a count — see eventlog.Record.
	Rev   uint64 `json:"rev,omitempty"`
	DurMs int64  `json:"durMs,omitempty"`
}

// EventList is the `events` result.
type EventList struct {
	Events []EventRow `json:"events"`
	// Files names every segment the read covered. "No events" over four
	// files is a different fact from "no events" over none, and only the
	// second one means nothing has ever been recorded.
	Files []string `json:"files,omitempty"`
	// Skipped counts undecodable lines.
	Skipped int `json:"skipped,omitempty"`
	// Note carries the "nothing here yet" diagnosis, so a --json consumer
	// sees the same explanation a human does.
	Note string `json:"note,omitempty"`
}

// Human renders the table.
func (l EventList) Human(w io.Writer) error {
	if len(l.Events) == 0 {
		if l.Note != "" {
			_, err := fmt.Fprintln(w, l.Note)
			return err
		}
		_, err := fmt.Fprintln(w, "no events recorded yet")
		return err
	}
	tw := tabwriter.NewWriter(w, 0, 4, 2, ' ', 0)
	_, _ = fmt.Fprintln(tw, "TIME\tSCOPE\tKIND\tSUBJECT\tDETAIL")
	for _, e := range l.Events {
		_, _ = fmt.Fprintf(tw, "%s\t%s\t%s\t%s\t%s\n",
			e.TS, e.Scope, e.Kind, dash(eventSubject(e)),
			oneLine(eventDetail(e), eventsDetailWidth))
	}
	if err := tw.Flush(); err != nil {
		return err
	}
	if l.Skipped > 0 {
		_, err := fmt.Fprintf(w, "%d undecodable line(s) skipped\n", l.Skipped)
		return err
	}
	return nil
}

// eventSubject is what the record is ABOUT: a server (with its derived
// instance when there is one), or the client whose gateway spoke.
func eventSubject(e EventRow) string {
	if e.Server == "" {
		return e.Client
	}
	if e.Inst != "" {
		return e.Server + "/" + e.Inst
	}
	return e.Server
}

// eventDetail folds the transition, the count and the free text into one
// column, in that order — a flip is the fact, and the prose elaborates it.
func eventDetail(e EventRow) string {
	var parts []string
	if e.From != "" || e.To != "" {
		parts = append(parts, dash(e.From)+"->"+dash(e.To))
	}
	if e.Count != 0 {
		parts = append(parts, countPhrase(e.Kind, e.Count))
	}
	if e.Rev != 0 {
		parts = append(parts, fmt.Sprintf("rev %d", e.Rev))
	}
	if e.DurMs != 0 {
		parts = append(parts, fmt.Sprintf("%dms", e.DurMs))
	}
	if e.Detail != "" {
		parts = append(parts, e.Detail)
	}
	return strings.Join(parts, " ")
}

// countPhrase renders Count with the noun its kind gives it.
//
// The fallback is a bare `n=13`, and it is what a kind this build does not
// know gets — a record from a newer daemon still prints its number rather
// than losing it. That is also what every count read as before the noun
// existed, which is why a connect that listed thirteen tools looked like a
// thirteenth attempt.
func countPhrase(kind string, n int) string {
	if noun := eventlog.CountNoun(eventlog.Kind(kind)); noun != "" {
		return fmt.Sprintf("%d %s", n, noun)
	}
	return fmt.Sprintf("n=%d", n)
}

func (a *App) newEventsCmd() *cobra.Command {
	var (
		follow bool
		since  time.Duration
		limit  int
		scope  string
		server string
		client string
		kinds  []string
	)
	cmd := &cobra.Command{
		Use:   "events [--server <id>] [--since 24h] [-f]",
		Short: "Read server, gateway and daemon state changes",
		Long: "Reads <data>/logs/events.jsonl — every state change of a downstream\n" +
			"server, a gateway or the daemon, in a closed vocabulary.\n\n" +
			"Scopes: " + strings.Join(eventlog.ScopeNames(), ", ") + ".\n" +
			"Works offline: a stdio gateway writes this file with no daemon running.\n\n" +
			"See also `agenthub logs` for the same processes' prose, and\n" +
			"`agenthub audit` for what a client CALLED.",
		Args: noArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			q, err := eventQuery(scope, server, client, kinds, since)
			if err != nil {
				return err
			}
			logsDir, err := a.resolver.LogsDir()
			if err != nil {
				return err
			}
			path := filepath.Join(logsDir, eventlog.FileName)
			if follow {
				return a.followEvents(cmd.Context(), path, q, limit)
			}
			list, err := readEventList(path, q, limit)
			if err != nil {
				return err
			}
			return a.printer().Emit(list)
		},
	}
	cmd.Flags().BoolVarP(&follow, "follow", "f", false, "stay open and keep printing new events")
	cmd.Flags().DurationVar(&since, "since", 0, "only events newer than this age (e.g. 24h)")
	cmd.Flags().IntVar(&limit, "limit", eventsDefaultLimit, "how many events to show (0 = all of them)")
	cmd.Flags().StringVar(&scope, "scope", "", "only this scope: "+strings.Join(eventlog.ScopeNames(), ", "))
	cmd.Flags().StringVar(&server, "server", "", "only events about this downstream server")
	cmd.Flags().StringVar(&client, "client", "", "only events observed by this client's gateway")
	cmd.Flags().StringSliceVar(&kinds, "kind", nil, "only these kinds (comma separated)")
	return cmd
}

// eventQuery validates the selectors and builds the read query.
//
// An unknown scope or kind is a USAGE ERROR, not an empty result. The
// vocabulary is closed precisely so a caller can be told it got a name
// wrong; answering a typo with "no events" hands back the same output as
// "this has not happened", which is the one confusion a closed set exists to
// prevent.
//
// Both the check and the hint come from internal/eventlog. This command and
// the control plane's /events/log ask the same question of the same closed
// set, and a local copy of the answer is one that can be right while the
// other is wrong — which is how the control plane ended up hinting a list of
// scopes written by hand.
func eventQuery(scope, server, client string, kinds []string, since time.Duration) (eventlog.Query, error) {
	q := eventlog.Query{Server: server, Client: client}
	if since > 0 {
		q.Since = time.Now().Add(-since)
	}
	if scope != "" {
		if !eventlog.KnownScope(eventlog.Scope(scope)) {
			e := Usagef("unknown scope %q", scope)
			e.Hint = "known scopes: " + strings.Join(eventlog.ScopeNames(), ", ")
			return eventlog.Query{}, e
		}
		q.Scope = eventlog.Scope(scope)
	}
	for _, raw := range kinds {
		k := eventlog.Kind(strings.TrimSpace(raw))
		if !eventlog.KnownKind(q.Scope, k) {
			e := Usagef("unknown kind %q", raw)
			e.Hint = "known kinds: " + strings.Join(eventlog.KindNames(q.Scope), ", ")
			return eventlog.Query{}, e
		}
		q.Kinds = append(q.Kinds, k)
	}
	return q, nil
}

// readEventList reads and projects, keeping the newest `limit` records.
func readEventList(path string, q eventlog.Query, limit int) (EventList, error) {
	res, err := eventlog.Read(path, q)
	if err != nil {
		return EventList{}, err
	}
	records := res.Records
	if limit > 0 && len(records) > limit {
		records = records[len(records)-limit:]
	}
	list := EventList{Events: rowsOfEvents(records), Files: res.Files, Skipped: res.Skipped}
	if len(res.Files) == 0 {
		list.Note = fmt.Sprintf("no event stream at %s yet; any running gateway or daemon "+
			"writes it unless 'events.enabled' is false", path)
	}
	return list, nil
}

func rowsOfEvents(records []eventlog.Record) []EventRow {
	out := make([]EventRow, 0, len(records))
	for _, r := range records {
		out = append(out, EventRow{
			TS: r.TS.UTC().Format(time.RFC3339), Scope: string(r.Scope), Kind: string(r.Kind),
			Server: r.Server, Inst: r.Inst, Client: r.Client, PID: r.PID,
			From: r.From, To: r.To, Detail: r.Detail,
			Count: r.Count, Rev: r.Rev, DurMs: r.DurMs,
		})
	}
	return out
}

// followEvents prints the current tail, then every newly appended record.
//
// It tracks a record TIMESTAMP rather than a byte offset, unlike the other
// two followers. This stream rotates into segments that Read walks as one
// sequence, so a byte offset into "the file" is not a position in the stream
// at all: after a rotation it points into a segment that is no longer the
// end.
func (a *App) followEvents(ctx context.Context, path string, q eventlog.Query, limit int) error {
	p := a.printer()
	list, err := readEventList(path, q, limit)
	if err != nil {
		return err
	}
	if err := p.Emit(list); err != nil {
		return err
	}
	seen := lastSeen(list.Events)
	ticker := time.NewTicker(eventsFollowInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return nil
		case <-ticker.C:
		}
		res, err := eventlog.Read(path, q)
		if err != nil {
			return err
		}
		fresh := rowsOfEvents(res.Records)
		if !seen.IsZero() {
			fresh = after(fresh, seen)
		}
		if len(fresh) == 0 {
			continue
		}
		seen = lastSeen(fresh)
		if err := p.Emit(EventList{Events: fresh}); err != nil {
			return err
		}
	}
}

// lastSeen is the timestamp of the newest row, or the zero time.
func lastSeen(rows []EventRow) time.Time {
	if len(rows) == 0 {
		return time.Time{}
	}
	ts, err := time.Parse(time.RFC3339, rows[len(rows)-1].TS)
	if err != nil {
		return time.Time{}
	}
	return ts
}

// after keeps rows strictly newer than ts.
//
// Strictly, so a record sharing the last-printed second is dropped rather
// than printed twice. That is a real trade and is stated because it is one:
// the rendered stamp has second resolution, so a burst inside one second can
// lose its tail here. A duplicate is the worse failure in a stream someone
// is watching for a state change — it reads as the state having changed
// twice.
func after(rows []EventRow, ts time.Time) []EventRow {
	for i, r := range rows {
		t, err := time.Parse(time.RFC3339, r.TS)
		if err != nil {
			continue
		}
		if t.After(ts) {
			return rows[i:]
		}
	}
	return nil
}
